src/Functions/parseDateTime.cpp

```diff
@@ -684,57 +694,116 @@ namespace
+            if constexpr (return_type == ReturnType::DateTime64 && parse_syntax == ParseSyntax::Joda)
+            {
+                UInt32 s_count = 0;
+                for (const auto & instruction : instructions)
+                {
+                    const String & fragment = instruction.getFragment();
+                    for (char ch : fragment)
+                    {
+                        if (ch != 'S')
+                        {
+                            s_count = 0;
+                            break;
+                        }
+                        else
+                            ++s_count;
+                    }
+                }
+                /// If the number of 'S' characters in format string not euqals the scale, then throw an exception to report error.
+                if (s_count != 0 && s_count != scale)
+                    result = tl::unexpected(ErrorCodeAndMessage(ErrorCodes::BAD_ARGUMENTS,
+                        "The number of 'S' characters in the input format string {} does not equal the given scale value {}.",
+                        format,
+                        std::to_string(scale)));
+            }
+
+            const auto & time_zone = getTimeZone(arguments);
             /// Make datetime fit in a cache line.
             alignas(64) DateTime<error_handling> datetime;
             for (size_t i = 0; i < input_rows_count; ++i)
             {
                 datetime.reset();
+
                 StringRef str_ref = col_str->getDataAt(i);
                 Pos cur = str_ref.data;
                 Pos end = str_ref.data + str_ref.size;
                 bool error = false;
 
+                auto handle_error = [&]([[maybe_unused]] const auto & res) -> void
+                {
+                    if constexpr (error_handling == ErrorHandling::Zero)
+                    {
+                        res_data[i] = 0;
+                        error = true;
+                    }
+                    else if constexpr (error_handling == ErrorHandling::Null)
+                    {
+                        res_data[i] = 0;
+                        col_null_map->getData()[i] = 1;
+                        error = true;
+                    }
+                    else
+                    {
+                        static_assert(error_handling == ErrorHandling::Exception);
+                        const ErrorCodeAndMessage & err = res.error();
+                        throw Exception(err.error_code, "{}", err.error_message);
+                    }
+                };
+
+                if (!result.has_value())
+                {
+                    handle_error(result);
+                    continue;
+                }
+
+                if constexpr (return_type == ReturnType::DateTime64)
+                {
+                    if (auto res = datetime.setScale(static_cast<UInt8>(scale), parse_syntax); !res.has_value())
+                    {
+                        handle_error(res);
```

Comment:
I think calling `handle_error` here is overkill. We only end up here if the scale argument is badly incorrect. Note that the scale argument is const anyways. I think we should unconditionally throw an exception here and not call `handle_error`.

Similarily, l. 774-778 are new. I don't understand what they are for? Is this to handle the error generated in l. 735? The same argument applies here, we should instead unconditionally throw an exception in l. 735 and remove l. 774-778.

Once that is done, only a single call to `handle_error` is left (the one in l. 797). This means we can afford to move the error handling code back (like it was before) instead of factorizing it into a lambda function (l. 753-772).